File: app/src/main/java/uk/co/sundroid/activity/data/fragments/dialogs/date/MonthPickerFragment.java
package uk.co.sundroid.activity.data.fragments.dialogs.date;

import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.app.Fragment;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.View;
import android.widget.NumberPicker;
import uk.co.sundroid.R.id;
import uk.co.sundroid.R.layout;

import java.util.Calendar;

import static java.util.Calendar.*;

public class MonthPickerFragment extends DialogFragment implements DialogInterface.OnClickListener {

    private int year;
    private int month;
    private NumberPicker monthPicker;
    private NumberPicker yearPicker;

    @FunctionalInterface
    public interface OnMonthSelectedListener {
        void onMonthSet(int year, int month);
    }

    public static MonthPickerFragment newInstance(Calendar calendar) {
        MonthPickerFragment fragment = new MonthPickerFragment();
        Bundle args = new Bundle();
        args.putInt("y", calendar.get(YEAR));
        args.putInt("m", calendar.get(MONTH));
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Calendar today = getInstance();
        this.year = getArguments().getInt("y", today.get(YEAR));
        this.month = getArguments().getInt("m", today.get(MONTH));
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        setRetainInstance(true); // TODO may not be required
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());

        View content = View.inflate(getActivity(), layout.dialog_monthpicker, null);
        monthPicker = content.findViewById(id.monthPicker);
        monthPicker.setMinValue(0);
        monthPicker.setMaxValue(11);
        monthPicker.setWrapSelectorWheel(true);
        monthPicker.setDisplayedValues(new String[] { "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec" });
        monthPicker.setValue(month);
        yearPicker = content.findViewById(id.yearPicker);
        yearPicker.setMinValue(1900);
        yearPicker.setMaxValue(2100);
        yearPicker.setWrapSelectorWheel(false);
        yearPicker.setValue(year);

        builder.setView(content);
        builder.setTitle("Set month");
        builder.setPositiveButton("Set", this);
        builder.setNeutralButton("This month", this);
        builder.setNegativeButton("Cancel", this);
        return builder.create();
    }

    @Override
    public void onClick(DialogInterface dialogInterface, int button) {
        Fragment target = getTargetFragment();
        if (target != null && target instanceof OnMonthSelectedListener) {
            if (button == DialogInterface.BUTTON_NEUTRAL) {
                Calendar today = getInstance();
                ((OnMonthSelectedListener)target).onMonthSet(today.get(YEAR), today.get(MONTH));
            } else if (button == DialogInterface.BUTTON_POSITIVE) {
                ((OnMonthSelectedListener)target).onMonthSet(yearPicker.getValue(), monthPicker.getValue());
            }
        }
        dismiss();
    }

}

File: app/src/main/kotlin/uk/co/sundroid/activity/data/fragments/dialogs/date/MonthPickerFragment.kt
package uk.co.sundroid.activity.data.fragments.dialogs.date

import android.app.AlertDialog
import android.app.Dialog
import android.os.Bundle
import androidx.fragment.app.DialogFragment
import uk.co.sundroid.activity.data.fragments.AbstractMonthFragment
import uk.co.sundroid.databinding.DialogMonthpickerBinding
import uk.co.sundroid.util.time.shortMonths
import java.util.*
import java.util.Calendar.*

class MonthPickerFragment : DialogFragment() {

    private var calendar: Calendar = getInstance()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        restore(arguments)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        save(calendar, outState)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        restore(savedInstanceState)

        val view = DialogMonthpickerBinding.inflate(layoutInflater)

        view.monthPicker.apply {
            minValue = 0
            maxValue = 11
            wrapSelectorWheel = true
            displayedValues = shortMonths
            value = calendar.get(MONTH)
            setOnValueChangedListener { _, _, month -> calendar.set(MONTH, month) }
        }
        view.yearPicker.apply {
            minValue = 1900
            maxValue = 2100
            wrapSelectorWheel = false
            value = calendar.get(YEAR)
            setOnValueChangedListener { _, _, year -> calendar.set(YEAR, year) }
        }

        return AlertDialog.Builder(activity).apply {
            setView(view.root)
            setTitle("Set month")
            setPositiveButton("Set") { _, _ -> set(calendar) }
            setNeutralButton("This month") { _, _ -> set(getInstance()) }
            setNegativeButton("Cancel", null)
        }.create()
    }

    private fun set(calendar: Calendar) {
        val target = targetFragment
        if (target is AbstractMonthFragment<*>) {
            target.calendarSet(calendar.get(YEAR), calendar.get(MONTH), target.getDateCalendar().get(DAY_OF_MONTH))
        }
    }

    private fun restore(bundle: Bundle?) {
        if (bundle != null) {
            val ymd = bundle.getIntArray("ymd")
            val tz = bundle.getString("tz")
            if (ymd != null && tz != null) {
                calendar.timeZone = TimeZone.getTimeZone(tz)
                calendar.set(ymd[0], ymd[1], ymd[2])
            }
        }
    }

    companion object {
        fun newInstance(calendar: Calendar): MonthPickerFragment {
            return MonthPickerFragment().apply {
                arguments = save(calendar)
            }
        }

        private fun save(calendar: Calendar, bundle: Bundle? = Bundle()): Bundle? {
            return bundle?.apply {
                putIntArray("ymd", intArrayOf(calendar.get(YEAR), calendar.get(MONTH), calendar.get(DAY_OF_MONTH)))
                putString("tz", calendar.timeZone.id)
            }
        }
    }

}